dipy/reconst/beltrami.py

```diff
@@ -0,0 +1,393 @@
+    beta : float
+        The ratio between the the spatial feature manifold and
+        the image domain metrics, controls how isottropic is the smoothing of the
+        D manifold, see [2] for more details.
+
+    Returns
+    -------
+    bad_g : boolean array
+        Boolean mask that marks the indices of the data where the induced metric
+        tensor "g" is unstable (i.e. the determinant of g is negative or increases to
+        extreme values), and where the update should be null.
+    gsqr : ndarray
+        The square root of the determinant of the metric tensor g.
+    dB : (..., 6) ndarray
+        The increments to be applied to each component of D, computed by applying
+        the Beltrami operator.
+
+    Notes
+    -----
+    In the Beltrami framework for DTI, a 3 by 3 symmetric tensor $g$ is used
+    to measure distances on the 3D image domain, defined (in Einstein summation) by:
+
+        $g_{ij} = \partial_{i} x^{m} \partial_{j} x^{n} h_{mn}$,
+
+    where the $$x^{m}$ denotes the "mth" component of a given manifold
+    (e.g. Dxx of D), the $\partial_{i}$ denotes the derivative along the
+    "ith" axis and $h_{mn}$ is the spatial feature metric tensor.
+    However, if this metric is chosen to be euclidean, then there is no need
+    for the Iwasawa conversion and $h$ becomes a 9 by 9 diagonal matrix, with
+    the first 3 diagonal elements equal to 1 and the rest equal to the constant
+    beta [1].
+
+    The Beltrami operator for the "nth" component of a given manifold $x^{n}$
+    is defined in terms of the inverse of $g$:
+
+        $\Delta B^{n} = 
+        /frac{1}{\sqrt{|g|}} \partial_{i} \sqrt{|g|} g^{ij} \partial_{j} x^{n}$,
+    
+    where $g|$ denotes the determinant of $g$ and  $g^{ij}$  its inverse.
+    
+    Applying this operator refularizes (smooths) the manifold D, small "beta"
+    ratios result in more isotropic smoothing. See references for more details.
+
+    References
+    ----------
+    .. [1] Pasternak, O., Maier-Hein, K., Baumgartner,
+        C., Shenton, M. E., Rathi, Y., & Westin, C. F. (2014).
+        The estimation of free-water corrected diffusion tensors.
+        In Visualization and Processing of Tensors and Higher Order
+        Descriptors for Multi-Valued Data (pp. 249-270). Springer,
+        Berlin, Heidelberg.
+
+    .. [2] Gur, Y., & Sochen, N. (2007, October).
+        Fast invariant Riemannian DT-MRI regularization.
+        In 2007 IEEE 11th International Conference on Computer Vision (pp. 1-7). IEEE.
+
+    """
+
+    nx, ny, nz = D.shape[:-1]
+    dx, dy, dz = zooms
+
+    Dxx = D[..., 0]
+    Dxy = D[..., 1]
+    Dyy = D[..., 2]
+    Dxz = D[..., 3]
+    Dyz = D[..., 4]
+    Dzz = D[..., 5]
+
+    # computing all spatial derivatives of D,
+    # Xi denotes the "ith" components of D (e.g. X1x = d(Dxx) / dx)
+    X1x = forward_diff(Dxx, 0, dx) * mask  # not sure if it is efficient to constantly multiply by this mask,
```

Comment:
To avoid multiplying constantly the variables by mask, can just mask once the inputs of this function?

Reply:
Will the border derivatives effect a lot the performance of the algorithm?

Reply:
If I mask the Iwasawa array, it becomes linearized and I loose the spatial information to make the derivatives, unless I use some equivalent function of Matlab's "ind2sub/sub2ind" to store which indices in the linearized array correspond to the shifted indices (n+1) along all dimensions of the 3D array, but those extra operations might end being more costly.

Also, I thought masking was efficient because it was creating a "view" of the original data, like slicing with (:), bu recently I discovered that masking actually makes copies, which is less efficient than creating a view. So multiplying by this mask is not so bad as I thought, at least for the array size that I'm working with.

If the border derivatives are not dealt with, it might lead to unstable "g" metrics in those voxels, that can propagate to neighboring voxels with successive iteartions, however I already have a guard in the code to avoid updating voxels with unstable metrics, so this masking might not be absolutely necessary, but it is always good to have an extra precaution.